File: beecrowd/compilador.py
def verificar_compilacao(programa):
    # Dicionário para rastrear as variáveis que foram atribuídas
    variaveis_atribuidas = {}
    # Conjunto para rastrear as variáveis usadas como fonte
    variaveis_fonte = set()
    
    for linha in programa:
        # Pula linhas vazias
        if not linha.strip():
            continue
            
        tokens = linha.strip().split()
        
        # Verifica se a variável já foi atribuída anteriormente
        var_destino = tokens[0]
        if var_destino in variaveis_atribuidas:
            return "Compilation Error"  # Variável só pode ser atribuída uma vez
        variaveis_atribuidas[var_destino] = True
        
        if len(tokens) == 3:  # Instrução de atribuição: A := B
            if not tokens[2].isdigit():  # Se B não for um número, deve ser uma variável
                if tokens[2] not in variaveis_atribuidas:
                    return "Compilation Error"  # Uso de variável não inicializada
                variaveis_fonte.add(tokens[2])  # Adiciona à lista de variáveis fonte
                
        elif len(tokens) == 5:  # Instrução de soma: A := B + C
            # Verifica se B é uma variável e está inicializada
            if not tokens[2].isdigit():
                if tokens[2] not in variaveis_atribuidas:
                    return "Compilation Error"  # Uso de variável não inicializada
                variaveis_fonte.add(tokens[2])  # Adiciona à lista de variáveis fonte
                
            # Verifica se C é uma variável e está inicializada
            if not tokens[4].isdigit():
                if tokens[4] not in variaveis_atribuidas:
                    return "Compilation Error"  # Uso de variável não inicializada
                variaveis_fonte.add(tokens[4])  # Adiciona à lista de variáveis fonte
        
        # Verifica se o número de variáveis fonte excede 2
        if len(variaveis_fonte) > 2:
            return "Compilation Error"
            
    return "OK"

# Função principal
def principal():
    programa = []
    try:
        while True:
            linha = input()
            programa.append(linha)
    except EOFError:
        pass
    
    print(verificar_compilacao(programa))

if __name__ == "__main__":
    principal()

File: beecrowd/compilador.cpp
#include <algorithm>
#include <iostream>
#include <sstream>
#include <string>
#include <unordered_map>
#include <vector>
using namespace std;

struct InfoVar {
  int linhaDefinicao; // Linha em que a variável foi definida
  int ultimoUso;      // Última linha em que a variável é usada como operando
  bool utilizada; // Indica se a variável aparece como fonte (operando)
};

// Verifica se o token é um número (constante)
bool ehNumero(const string &s) {
  for (char c : s) if (!isdigit(c)) return false;
  return true;
}

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  unordered_map<string, InfoVar> variaveis;
  vector<string> linhas;
  string linha;
  while (getline(cin, linha)) {
    if (linha.empty()) continue; // Ignora linhas vazias
    linhas.push_back(linha);
  }

  int n = linhas.size();

  // Processa cada instrução
  for (int i = 0; i < n; i++) {
    istringstream fluxoEntrada(linhas[i]);
    vector<string> tokens;
    string token;
    while (fluxoEntrada >> token) tokens.push_back(token);
  
    // Instrução de declaração: "A := B" (3 tokens)
    if (tokens.size() == 3) {
      string variavel = tokens[0];
      // Se a variável já foi definida, erro
      if (variaveis.find(variavel) != variaveis.end()) {
        cout << "Compilation error";
        return 0;
      }
      // O token B deve ser um número
      if (!ehNumero(tokens[2])) {
        cout << "Compilation error";
        return 0;
      }
      InfoVar informacao;
      informacao.linhaDefinicao = i + 1;
      informacao.ultimoUso = -1; // Ainda não foi usada como operando
      informacao.utilizada = false;
      variaveis[variavel] = informacao;
    }
    // Instrução de soma: "A := B + C" (5 tokens)
    else if (tokens.size() == 5) {
      string variavel = tokens[0];
      // Verifica atribuição única
      if (variaveis.find(variavel) != variaveis.end()) {
        cout << "Compilation error";
        return 0;
      }
      // Para cada operando (tokens[2] e tokens[4])
      for (int j = 2; j <= 4; j += 2) {
        // Se o operando não for um número, deve ser variável já definida
        if (!ehNumero(tokens[j])) {
          if (variaveis.find(tokens[j]) == variaveis.end()) {
            cout << "Compilation error";
            return 0;
          }
          // Atualiza o último uso da variável para a linha atual
          InfoVar &informacao = variaveis[tokens[j]];
          informacao.utilizada = true;
          informacao.ultimoUso = max(informacao.ultimoUso, i + 1);
        }
      }
      // Registra a variável de destino
      InfoVar informacao;
      informacao.linhaDefinicao = i + 1;
      informacao.ultimoUso = -1;
      informacao.utilizada = false;
      variaveis[variavel] = informacao;
    } else {
      // Formato inesperado
      cout << "Compilation error";
      return 0;
    }
  }

  // Cria os intervalos de vida para as variáveis que foram usadas como operando
  // O intervalo é [linhaDefinicao, ultimoUso]
  vector<pair<int, int>> intervalos;
  for (auto &p : variaveis) {
    InfoVar informacao = p.second;
    if (informacao.utilizada) intervalos.push_back({informacao.linhaDefinicao, informacao.ultimoUso});
  }

  // Usa sweep‑line para contar o número máximo de intervalos sobrepostos
  vector<pair<int, int>> eventos;
  for (auto &intervalo : intervalos) {
    eventos.push_back({intervalo.first, 1});
    eventos.push_back({intervalo.second + 1, -1});
  }
  sort(eventos.begin(), eventos.end());
  int atual = 0, maxVivas = 0;
  for (auto &ev : eventos) {
    atual += ev.second;
    maxVivas = max(maxVivas, atual);
  }
  // Se em algum ponto mais de 2 variáveis estão vivas, não compila
  if (maxVivas > 2) cout << "Compilation error";
  else cout << "OK";

  return 0;
}
